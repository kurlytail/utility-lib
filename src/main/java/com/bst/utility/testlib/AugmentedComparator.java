package com.bst.utility.testlib;

import java.util.Comparator;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.NumericNode;

// This class is being used but is not doing anything, added here because it will be needed
// in the future
public class AugmentedComparator implements Comparator<JsonNode> {
	@Override
	public int compare(final JsonNode o1, final JsonNode o2) {
		if (o1.equals(o2)) {
			return 0;
		}
		if ((o1 instanceof NumericNode) && (o2 instanceof NumericNode)) {
			final double d1 = ((NumericNode) o1).asDouble();
			final double d2 = ((NumericNode) o2).asDouble();
			if (d1 == d2) { // strictly equals because it's integral value
				return 1;
			}
		}
		return 0;
	}
}
