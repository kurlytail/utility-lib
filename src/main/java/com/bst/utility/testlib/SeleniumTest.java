package com.bst.utility.testlib;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Inherited;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.safari.SafariDriver;

@Documented
@Inherited
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface SeleniumTest {

	String baseUrl() default "http://localhost:8080";

	Class<? extends WebDriver> driver() default SafariDriver.class;
}