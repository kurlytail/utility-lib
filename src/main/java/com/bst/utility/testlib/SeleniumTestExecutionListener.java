package com.bst.utility.testlib;

import org.openqa.selenium.WebDriver;
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.core.annotation.AnnotationUtils;
import org.springframework.test.context.TestContext;
import org.springframework.test.context.support.DependencyInjectionTestExecutionListener;

public class SeleniumTestExecutionListener extends DependencyInjectionTestExecutionListener {

	private WebDriver driver = null;

	@Override
	public void afterTestClass(final TestContext testContext) throws Exception {
		if (this.driver != null) {
			this.driver.close();
		}
	}

	@Override
	public void beforeTestClass(final TestContext testContext) throws Exception {
		final ApplicationContext context = testContext.getApplicationContext();
		if ((context instanceof ConfigurableApplicationContext) && (this.driver == null)) {
			final SeleniumTest annotation = AnnotationUtils.findAnnotation(testContext.getTestClass(),
					SeleniumTest.class);
			final ConfigurableListableBeanFactory beanFactory = ((ConfigurableApplicationContext) context)
					.getBeanFactory();
			this.driver = beanFactory.createBean(annotation.driver());
			beanFactory.registerSingleton("webDriver", this.driver);
		}
	}
}
