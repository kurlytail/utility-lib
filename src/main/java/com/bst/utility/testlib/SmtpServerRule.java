package com.bst.utility.testlib;

import javax.mail.internet.MimeMessage;

import org.junit.rules.ExternalResource;

import com.icegreen.greenmail.util.GreenMail;
import com.icegreen.greenmail.util.ServerSetup;

public class SmtpServerRule extends ExternalResource {

	private final int port;
	private GreenMail smtpServer;

	public SmtpServerRule(final int port) {
		this.port = port;
	}

	@Override
	protected void after() {
		super.after();
		this.smtpServer.stop();
	}

	@Override
	protected void before() throws Throwable {
		super.before();
		this.smtpServer = new GreenMail(new ServerSetup(this.port, null, "smtp"));
		this.smtpServer.start();
	}

	public MimeMessage[] getMessages() {
		return this.smtpServer.getReceivedMessages();
	}
}