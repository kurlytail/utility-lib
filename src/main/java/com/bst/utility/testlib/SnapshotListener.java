package com.bst.utility.testlib;

import org.springframework.test.context.TestContext;
import org.springframework.test.context.support.AbstractTestExecutionListener;

public class SnapshotListener extends AbstractTestExecutionListener {

	private static SnapshotMatcher matcher;

	public static SnapshotMatcher expect(final Object obj) {
		return SnapshotListener.matcher.matchObjectSnapshot(obj);
	}

	@Override
	public void afterTestClass(final TestContext testContext) throws Exception {
		SnapshotListener.matcher.commitSnapshot(testContext);
	}

	@Override
	public void afterTestMethod(final TestContext testContext) throws Exception {
		SnapshotListener.matcher.commitTest(testContext);
	}

	@Override
	public void beforeTestClass(final TestContext testContext) throws Exception {
		SnapshotListener.matcher = new SnapshotMatcher(testContext);
	}

	@Override
	public void beforeTestMethod(final TestContext testContext) throws Exception {
		SnapshotListener.matcher.startTest(testContext);
	}
}
