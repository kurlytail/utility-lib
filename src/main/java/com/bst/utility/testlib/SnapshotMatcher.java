package com.bst.utility.testlib;

import java.io.File;
import java.io.PrintWriter;

import org.springframework.test.context.TestContext;

import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class SnapshotMatcher {
	private JsonNode snapshots;
	private Object objectToMatch;
	private boolean snapshotsUpdated = false;
	private int snapshotSequence = -1;
	private String snapshotTestName;
	private ObjectMapper objectMapper = new ObjectMapper();

	SnapshotMatcher(TestContext testContext) throws Exception {
		objectMapper.setSerializationInclusion(Include.NON_NULL);
		String snapshotName = "src/test/java/" + testContext.getTestClass().getName().replaceAll("\\.", "/") + ".snap";
		File file = new File(snapshotName);
		snapshots = file.exists() ? objectMapper.readTree(file) : objectMapper.createObjectNode();
	}

	public void toMatchSnapshot() throws Exception {
		++snapshotSequence;

		JsonNode objectTree = objectMapper.valueToTree(objectToMatch);
		if (objectToMatch == null) {
			objectTree = objectMapper.createObjectNode();
		}

		if (!snapshots.has(snapshotTestName)) {
			JsonNode snapshotArray = objectMapper.createArrayNode();
			((ArrayNode) snapshotArray).add(objectTree);
			((ObjectNode) snapshots).set(snapshotTestName, snapshotArray);
			snapshotsUpdated = true;
			return;
		}

		JsonNode snapshotArray = snapshots.get(snapshotTestName);

		if (!snapshotArray.has(snapshotSequence)) {
			snapshotArray = ((ArrayNode) snapshotArray).add(objectTree);
			snapshotsUpdated = true;
			return;
		}

		JsonNode compareTo = snapshotArray.get(snapshotSequence);
		if (objectTree == compareTo || !objectTree.toString().equals(compareTo.toString())) {
			snapshotsUpdated = false;
			throw new Exception("Snapshot mismatch for " + snapshotTestName + " at " + snapshotSequence);
		}
	}

	public void commitTest(TestContext testContext) {
	}

	public void commitSnapshot(TestContext testContext) throws Exception {
		if (snapshotsUpdated) {
			String snapshotName = "src/test/java/" + testContext.getTestClass().getName().replaceAll("\\.", "/")
					+ ".snap";
			File file = new File(snapshotName);

			PrintWriter outputFile = new PrintWriter(file.getAbsolutePath());
			String finalSnapshot = objectMapper.writerWithDefaultPrettyPrinter()
					.writeValueAsString(objectMapper.readTree(snapshots.toString()));
			outputFile.write(finalSnapshot);
			outputFile.close();
		}
	}

	public SnapshotMatcher matchObjectSnapshot(Object obj) {
		objectToMatch = obj;
		return this;
	}

	public void startTest(TestContext testContext) {
		snapshotSequence = -1;
		snapshotsUpdated = false;
		snapshotTestName = testContext.getTestMethod().getName();
	}
}
