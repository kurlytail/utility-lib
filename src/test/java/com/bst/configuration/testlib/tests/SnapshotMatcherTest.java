package com.bst.configuration.testlib.tests;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.springframework.test.context.TestContext;
import org.springframework.test.context.TestExecutionListeners;
import org.springframework.test.context.TestExecutionListeners.MergeMode;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import com.bst.utility.testlib.SnapshotListener;

@ExtendWith(SpringExtension.class)
@TestExecutionListeners(listeners = SnapshotListener.class, mergeMode = MergeMode.MERGE_WITH_DEFAULTS)
public class SnapshotMatcherTest {

	public class TestClass {
		public Long value = (long) 20;
	}

	@Mock
	private TestContext testContext;

	@Test
	public void testUseLong() throws Exception {
		SnapshotListener.expect(new TestClass()).toMatchSnapshot();
	}

}
