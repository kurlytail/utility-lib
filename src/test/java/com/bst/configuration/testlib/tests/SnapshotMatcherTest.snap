{
  "useLongTest" : [ {
    "value" : 10
  } ],
  "testUseLong" : [ {
    "value" : 10
  } ]
}